File: Backend/src/data/Generics/Generics.ts
import {   OkPacket, ProcedureCallPacket, ResultSetHeader, RowDataPacket } from "mysql2";
import { ConnectionMysql } from "../Connection/ConnectionMysql";
import { IGenerics } from "./Interfaces/IGenerics";
import { GenericResult } from "./Enums/GenericResult";



export class Generics<T> implements IGenerics<T>{
    async CheckPerProperty(tabela: string, property: string, value: any): Promise<boolean> {
        try{
            let valueString = "";
            if(typeof(value) === 'number'){
                valueString = `${value}`
            }
            if(typeof(value) === 'string'){
                valueString = `"${value}"`
            }
            const conn = new ConnectionMysql();
            let query = conn.VerifyValueByPropertyExists(tabela,property, valueString);
            const [result]:[any[], any] = await (await conn.connection).query(query);
            await conn.close();
            return result[0];
        }catch(err){
            throw new Error(`${err}`)
        }
    }
    async GetbyId(tabela: string, id: number): Promise<object> {
        const conn = new ConnectionMysql();
        let query = conn.QuerySelectById(tabela , id);
        await conn.init()
        const [row]:[any[], any] = await (await conn.connection).execute(query)
        await conn.close();
        return row[0]
    }
    async findBy( NameTable: string, property: string, value:any): Promise<object> {
        let result:any[]=[];
        let tabela = NameTable.toLowerCase();
        let valueString = "";
        if(typeof(value) === 'number'){
            valueString = `${value}`
        }
        if(typeof(value) === 'string'){
            valueString = `"${value}"`
        }
        try{
            const conn = new ConnectionMysql();
            let query = conn.QuerySelectByProperty(tabela,property,valueString);
            await conn.init();
            const [result]:[any[], any] = await (await conn.connection).query(query);
            await conn.close();
            return result[0]
        }catch(err){
            console.error(`${err}`)
            return result[0]
        }
    }
    async save(entity: object): Promise<string> {
        let tabela = entity.constructor.name.toLowerCase();
        let valores: any[] = []
        let colunas: string[] = []
        const objCurrent = Object.assign(entity,entity)
        const properties = Object.getOwnPropertyNames(entity);
        properties.forEach((element:string, index:number)=>{
            if(element !== 'id'){
                colunas.push(element)
                let valorCurrent = Reflect.get(objCurrent,element)
                if(typeof(valorCurrent) === 'string'){
                    valores.push(`"${valorCurrent}"`)
                }
                if(typeof(valorCurrent) === 'number'){
                    valores.push(`${valorCurrent}`)
                }
            }
        })
        try{
            const conn = new ConnectionMysql();
            let query = conn.QueryInsert(tabela,colunas,valores);
            await conn.init();
            const [results] = await (await conn.connection).execute(query);
            console.log('Registro inserido com sucesso!');
            console.log('Resultado:', results);

            await conn.close();
            console.log('Query executada no banco de dados: ' + query);
            return GenericResult.Sucess;
        }catch(err){
            console.error(`${err}`)
            return GenericResult.Fail
        }

    }
    async listAll(tabela:string): Promise<Array<any>> {
        let rows: any[] = []
        try{
            const conn = new ConnectionMysql();
            let query = conn.QuerySelectAll(tabela);
            await conn.init();
            const [rows, fields]: [Array<any>, any] = await (await conn.connection).execute(query);
            console.log('Resultados:', rows);
            console.log('Campos:', fields);
            await conn.close();
            return rows
        }catch(err){
            console.error(`${err}`);
            return rows
        }
    }
    async update(entity:object): Promise<string> {
        let tabela = entity.constructor.name.toLowerCase();
        let referencia = "";
        let valoresEColunas: string[] = [];
        const properties = Object.getOwnPropertyNames(entity);
        properties.forEach((element:string, index:number)=>{
            if(element === 'id'){
                referencia = ` ${element} = ${Reflect.get(entity,element)}`
            }else{
                let valorCurrent = Reflect.get(entity,element)
                if(typeof(valorCurrent) === 'string'){
                    valoresEColunas.push(` ${element} = "${valorCurrent}"`)
                }
                if(typeof(valorCurrent) === 'number'){
                    valoresEColunas.push(`${element} = ${valorCurrent}`)
                }
            }

        })
        try{
            const conn = new ConnectionMysql();
            let query = conn.QueryUpdate(tabela,valoresEColunas, referencia)
            await conn.init();
            const [results] =  await (await conn.connection).execute(query);
            console.log('Registro atualizado com sucesso!');
            console.log('Resultado:', results);

            await conn.close();
            console.log('Query executada no banco de dados: ' + query);
            return GenericResult.Sucess;
        }catch(err){
            console.error(`${err}`)
            return GenericResult.Fail
        }

    }

}

File: Back/src/data/Generic/Generics.py
from typing import TypeVar, List
from mysql.connector import errors, Error
from data.Connection.ConnectionMysql import ConnectionMysql
T = TypeVar('T');

class BaseGenerics:
    def __init__(self) -> None:
        super().__init__()
    async def Save(self, value: T)->T:
        try:
            conn = ConnectionMysql();
            await conn.init();
            valores = [];
            campos = [];
            tabela = value.__class__.__name__.lower();
            colunas:str = ""
            values = ""
            propriedades = [propriedade for propriedade in vars(value)];
            for prop in propriedades:
                campos.append(prop);
                value_param = getattr(value, prop)
                if type(value_param) == str:
                    valores.append('"{}"'.format(value_param));
                elif type(value_param) == int:
                    valores.append(value_param)
                elif type(value_param) == bytes:
                    valor_byte = value_param.hex()
                    valores.append('"{}"'.format(valor_byte))
            for i in range(len(campos)):
                colunas += "{},".format(campos[i]);
                if type(valores[i]) == str:
                    string: str = valores[i]
                    values += '{}'.format(string)
                    values += ",";
                else:
                    values += "{},".format(valores[i]);
            query = conn.QueryInsert(tabela, colunas, values);
            await conn.executeAndCommit(query);
            await conn.close();
            return "Inserido no banco com sucesso"
        except Error as e:
            print("Erro no banco de dados:", e.msg)
    async def List(self, tabela:str):
        try:
            lista: list = [];
            conn = ConnectionMysql();
            await conn.init();
            query = conn.QuerySelectAll(tabela);
            await conn.execute(query);
            lista = conn.cursor.fetchall()
            await conn.close();
            return lista

        except Error as err:
            print("Erro no banco de dados:", err.msg)
    async def Update(self, value: T)-> T:
        try:
            conn = ConnectionMysql()
            await conn.init();
            colunas = [];
            valores = [];
            tabela = value.__class__.__name__.lower();
            referencia = ""
            identificador = "";
            valor_identificador = "";
            corpo = ""
            propriedades = [propriedade for propriedade in vars(value)];
            for prop in propriedades:
                if prop.lower() == 'id':
                    identificador = prop;
                    valor_identificador = getattr(value, prop);
                else:
                    colunas.append(prop);
                    value_param = getattr(value, prop);
                    if type(value_param) == str:
                        valores.append('"{}"'.format(value_param))

            for i in range(colunas.__len__()):
                corpo += ' {} = {},'.format(colunas[i], valores[i])

            referencia = '{} = {}'.format(identificador, valor_identificador)
            query = conn.QueryUpdate(tabela,corpo,referencia);
            await conn.executeAndCommit(query);
            await conn.close();
        except Error as e:
            print("Erro no banco de dados:", e.msg)
    async def FindById(self, tabela:str, id:int):
        try:
            conn = ConnectionMysql();
            await conn.init();
            query = conn.QueryFindById(tabela, id);
            await conn.execute(query);
            busca = conn.cursor.fetchall();
            await conn.close();
            return busca;
        except Error as err:
            return err;
    async def Delete(self, tabela:str, id:int)-> T:
        try:
            conn = ConnectionMysql();
            await conn.init();
            query = conn.QueryDeleteById(tabela, id);
            await conn.executeAndCommit(query);
            await conn.close();
        except Error as err:
            return err;
    async def CheckPerProperty(self, tabela: str, property:str, value:any):
        try:
            value_param: str;
            if(type(value) == str):
                value_param = '"{}"'.format(value);
            elif type(value) == int:
                value_param = '{}'.format(value);
            elif type(value) == float:
                value_param = '{}'.format(value);
            conn = ConnectionMysql();
            await conn.init();
            query = conn.VerifyValueByPropertyExists(tabela, property, value_param);
            await conn.execute(query);
            resultado = conn.cursor.fetchall();
            await conn.close();
            for item in resultado:
                if item['resultado'] == 'True':
                    return True
                return False


        except Error as err:
            print(err);
    async def FindBy(self, tabela:str, property:str, value:any):
        try:
            value_param: str;
            if(type(value) == str):
                value_param = '"{}"'.format(value);
            elif type(value) == int:
                value_param = '{}'.format(value);
            elif type(value) == float:
                value_param = '{}'.format(value);
            conn = ConnectionMysql();
            await conn.init();
            query = conn.QuerySelectByProperty(tabela, property, value_param);
            await conn.execute(query);
            resultado = conn.cursor.fetchall();
            await conn.close();
            for item in resultado:
                return item

        except Error as err:
            print(err)
    async def FindAllBy(self, tabela:str, property:str, value:any):
        try:
            lista : list = []
            value_param: str;
            if(type(value) == str):
                value_param = '"{}"'.format(value);
            elif type(value) == int:
                value_param = '{}'.format(value);
            elif type(value) == float:
                value_param = '{}'.format(value);
            conn = ConnectionMysql();
            await conn.init();
            query = conn.QuerySelectByProperty(tabela, property, value_param);
            await conn.execute(query);
            lista = conn.cursor.fetchall();
            await conn.close();
            return lista;
        except Error as err:
            print(err);
    async def CheckExistsEntityWithId(self, value: T):
        try:
            tabela = value.__class__.__name__.lower();
            id_column = "";
            other_column = "";
            propriedades = [propriedade for propriedade in vars(value)]
            for prop in propriedades:
                if prop.lower() == "id":
                    id_column = ' {} = {}'.format(prop, getattr(value,prop))
                else:
                    value_column = getattr(value, prop);
                    if type(value_column) == str:
                        other_column += ' and {} = "{}"'.format(prop, value_column);
                    else:
                        other_column += ' and {} = {}'.format(prop, value_column);
            values = '{} {}'.format(id_column, other_column);
            conn = ConnectionMysql();
            await conn.init();
            query = conn.VerifyEntity(tabela, values);
            await conn.execute(query);
            resultado = conn.cursor.fetchall();
            await conn.close();
            for item in resultado:
                if item['resultado'] == "True":
                    return True;
                else:
                    return False;
            return False;
        except Error as err:
            print(err);
    async def CheckExistsEntity(self, value:T):
        try:
            tabela = value.__class__.__name__.lower();
            referencies = "";
            propriedades = [propriedade for propriedade in vars(value)];
            for prop in propriedades:
                if prop.lower() != "id":
                    value_column = getattr(value, prop);
                    if type(value_column) == str:
                        referencies += ' {} = "{}",'.format(prop, getattr(value,prop));
                    else:
                        referencies += ' {} = {},'.format(prop, getattr(value,prop));
            referencies = referencies.rstrip(referencies[-1]);
            referencies = referencies.replace(',',' and');
            conn = ConnectionMysql();
            await conn.init();
            query = conn.VerifyValueByPropertiesExists(tabela,referencies)
            await conn.execute(query);
            resultado = conn.cursor.fetchall();
            await conn.close();
            for item in resultado:
                if item['resultado'] == "True":
                    return True;
                else:
                    return False;
            return False;

        except Error as err:
            print(err)
